File: kapor-backend/src/main/java/com/kapor/membyte/service/MembyteService.java
package com.kapor.membyte.service;

import com.kapor.common.exception.ResourceNotFoundException;
import com.kapor.devvocab.model.Lesson;
import com.kapor.devvocab.model.Topic;
import com.kapor.devvocab.repository.LessonRepository;
import com.kapor.devvocab.repository.TopicRepository;
import com.kapor.membyte.dto.MembyteDeckSummaryDto;
import com.kapor.membyte.dto.MembyteReviewCardDto;
import com.kapor.membyte.dto.MembyteReviewRatingRequest;
import com.kapor.membyte.dto.MembyteReviewResultDto;
import com.kapor.membyte.dto.MembyteReviewSummaryDto;
import com.kapor.membyte.dto.MembyteSaveCardDto;
import com.kapor.membyte.dto.MembyteSavedCardsDto;
import com.kapor.membyte.model.MembyteDeck;
import com.kapor.membyte.model.MembyteFlashcard;
import com.kapor.membyte.repository.MembyteDeckRepository;
import com.kapor.membyte.repository.MembyteFlashcardRepository;
import com.kapor.video.model.Video;
import com.kapor.video.repository.VideoRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class MembyteService {

    private final MembyteDeckRepository deckRepository;
    private final MembyteFlashcardRepository flashcardRepository;
    private final LessonRepository lessonRepository;
    private final TopicRepository topicRepository;
    private final VideoRepository videoRepository;
    private final FsrsService fsrsService;

    public MembyteSaveCardDto saveVocabulary(String userId, String lessonId, String vocabularyId) {
        Lesson lesson = lessonRepository.findById(lessonId)
                .orElseThrow(() -> new ResourceNotFoundException("Lesson", "id", lessonId));
        Lesson.VocabularyItem vocabulary = lesson.getVocabulary() == null ? null : lesson.getVocabulary().stream()
                .filter(item -> vocabularyId.equals(item.getId()))
                .findFirst()
                .orElse(null);
        if (vocabulary == null) {
            throw new ResourceNotFoundException("Vocabulary", "id", vocabularyId);
        }

        MembyteFlashcard existing = flashcardRepository
                .findByUserIdAndLessonIdAndVocabularyId(userId, lessonId, vocabularyId)
                .orElse(null);
        if (existing != null) {
            return MembyteSaveCardDto.builder()
                    .deckId(existing.getDeckId())
                    .cardId(existing.getId())
                    .vocabularyId(vocabularyId)
                    .alreadySaved(true)
                    .build();
        }

        Topic topic = topicRepository.findById(lesson.getTopicId())
                .orElseThrow(() -> new ResourceNotFoundException("Topic", "id", lesson.getTopicId()));
        Instant now = Instant.now();
        MembyteDeck deck = deckRepository.findByUserIdAndLessonId(userId, lessonId)
                .orElseGet(() -> deckRepository.save(MembyteDeck.builder()
                        .userId(userId)
                        .topicId(topic.getId())
                        .lessonId(lesson.getId())
                        .domain(topic.getDomain())
                        .title(joinLabels(topic.getTitle(), lesson.getTitle()))
                        .titleVi(joinLabels(topic.getTitleVi(), lesson.getTitleVi()))
                        .createdAt(now)
                        .updatedAt(now)
                        .build()));

        MembyteFlashcard saved = flashcardRepository.save(MembyteFlashcard.builder()
                .userId(userId)
                .deckId(deck.getId())
                .lessonId(lesson.getId())
                .vocabularyId(vocabulary.getId())
                .korean(vocabulary.getKorean())
                .pronunciation(vocabulary.getPronunciation())
                .vietnamese(vocabulary.getVietnamese())
                .english(vocabulary.getEnglish())
                .context(vocabulary.getContext())
                .codeSnippet(vocabulary.getCodeSnippet())
                .audioUrl(vocabulary.getAudioUrl())
                .isNew(true)
                .repetitions(0)
                .lapses(0)
                .difficulty(0)
                .stability(0)
                .createdAt(now)
                .build());
        deck.setUpdatedAt(now);
        deckRepository.save(deck);

        return MembyteSaveCardDto.builder()
                .deckId(deck.getId())
                .cardId(saved.getId())
                .vocabularyId(vocabularyId)
                .alreadySaved(false)
                .build();
    }

    public MembyteSaveCardDto saveVideoToken(String userId, String videoId, String surface) {
        Video video = videoRepository.findById(videoId)
                .orElseThrow(() -> new ResourceNotFoundException("Video", "id", videoId));
        VideoTokenMatch tokenMatch = findVideoToken(video, surface);
        if (tokenMatch == null) {
            throw new ResourceNotFoundException("Video token", "surface", surface);
        }

        // Reuse the existing lesson-based indexes with a namespaced source id.
        // This keeps one personal deck per video without a data migration.
        String videoSourceId = "video:" + videoId;
        Instant now = Instant.now();
        MembyteDeck deck = deckRepository.findByUserIdAndLessonId(userId, videoSourceId)
                .orElseGet(() -> deckRepository.save(MembyteDeck.builder()
                        .userId(userId)
                        .lessonId(videoSourceId)
                        .domain(video.getDomain())
                        .title(video.getTitle())
                        .titleVi(video.getTitleVi())
                        .createdAt(now)
                        .updatedAt(now)
                        .build()));

        Video.TokenizedWord token = tokenMatch.token();
        String savedWord = token.getStem() == null || token.getStem().isBlank()
                ? token.getSurface()
                : token.getStem();
        MembyteFlashcard existing = flashcardRepository.findByUserIdAndLessonId(userId, videoSourceId).stream()
                .filter(card -> savedWord.equals(card.getVocabularyId())
                        || token.getSurface().equals(card.getVocabularyId())
                        || savedWord.equals(card.getKorean()))
                .findFirst()
                .orElse(null);
        if (existing != null) {
            applyVideoTokenDetails(existing, token, tokenMatch.subtitleText(), savedWord);
            flashcardRepository.save(existing);
            return MembyteSaveCardDto.builder()
                    .deckId(existing.getDeckId())
                    .cardId(existing.getId())
                    .vocabularyId(savedWord)
                    .alreadySaved(true)
                    .build();
        }
        MembyteFlashcard saved = flashcardRepository.save(MembyteFlashcard.builder()
                .userId(userId)
                .deckId(deck.getId())
                .lessonId(videoSourceId)
                .vocabularyId(savedWord)
                .korean(savedWord)
                .pronunciation(token.getPronunciation())
                .vietnamese(token.getMeaningVi())
                .english(token.getMeaningEn())
                .definitionEn(token.getDefinitionEn())
                .exampleKo(token.getExampleKo())
                .grammarNote(token.getGrammarNote())
                .context(tokenMatch.subtitleText())
                .isNew(true)
                .repetitions(0)
                .lapses(0)
                .difficulty(0)
                .stability(0)
                .createdAt(now)
                .build());
        deck.setUpdatedAt(now);
        deckRepository.save(deck);

        return MembyteSaveCardDto.builder()
                .deckId(deck.getId())
                .cardId(saved.getId())
                .vocabularyId(savedWord)
                .alreadySaved(false)
                .build();
    }

    public MembyteSavedCardsDto getSavedVocabularyIds(String userId, String lessonId) {
        Set<String> vocabularyIds = flashcardRepository.findByUserIdAndLessonId(userId, lessonId).stream()
                .map(MembyteFlashcard::getVocabularyId)
                .collect(Collectors.toSet());
        return MembyteSavedCardsDto.builder().vocabularyIds(vocabularyIds).build();
    }

    public List<MembyteDeckSummaryDto> getDecks(String userId) {
        Instant now = Instant.now();
        return deckRepository.findByUserIdOrderByUpdatedAtDesc(userId).stream()
                .map(deck -> toDeckSummary(deck, flashcardRepository.findByUserIdAndDeckIdOrderByCreatedAtAsc(userId, deck.getId()), now))
                .toList();
    }

    public MembyteReviewSummaryDto getReviewSummary(String userId) {
        Instant now = Instant.now();
        Map<String, List<MembyteFlashcard>> cardsByDeck = flashcardRepository.findByUserIdOrderByCreatedAtAsc(userId).stream()
                .collect(Collectors.groupingBy(MembyteFlashcard::getDeckId));
        int totalDue = 0;
        int deckCount = 0;
        for (List<MembyteFlashcard> cards : cardsByDeck.values()) {
            int dueInDeck = (int) cards.stream().filter(card -> isDue(card, now)).count();
            totalDue += dueInDeck;
            if (dueInDeck > 0) deckCount++;
        }
        return MembyteReviewSummaryDto.builder()
                .totalDueCards(totalDue)
                .decksWithDueCards(deckCount)
                .build();
    }

    public List<MembyteReviewCardDto> getReviewCards(String userId, String deckId, boolean includeNew, int limit) {
        Instant now = Instant.now();
        List<MembyteFlashcard> cards;
        if (deckId == null || deckId.isBlank()) {
            cards = flashcardRepository.findByUserIdOrderByCreatedAtAsc(userId);
        } else {
            deckRepository.findById(deckId)
                    .filter(deck -> userId.equals(deck.getUserId()))
                    .orElseThrow(() -> new ResourceNotFoundException("Deck", "id", deckId));
            cards = flashcardRepository.findByUserIdAndDeckIdOrderByCreatedAtAsc(userId, deckId);
        }

        enrichVideoFlashcards(cards);

        return cards.stream()
                .filter(card -> isDue(card, now) || (includeNew && card.isNew()))
                .sorted(Comparator.comparing(MembyteFlashcard::isNew).thenComparing(MembyteFlashcard::getDueAt,
                        Comparator.nullsLast(Comparator.naturalOrder())))
                .limit(Math.max(1, Math.min(limit, 100)))
                .map(card -> toReviewCard(card, now))
                .toList();
    }

    public MembyteReviewResultDto rateCard(String userId, MembyteReviewRatingRequest request) {
        MembyteFlashcard card = flashcardRepository.findById(request.getCardId())
                .filter(savedCard -> userId.equals(savedCard.getUserId()))
                .orElseThrow(() -> new ResourceNotFoundException("Flashcard", "id", request.getCardId()));
        Instant now = Instant.now();
        FsrsService.Schedule schedule = fsrsService.rate(card, request.getRating(), now);
        flashcardRepository.save(card);
        deckRepository.findById(card.getDeckId()).ifPresent(deck -> {
            deck.setUpdatedAt(now);
            deckRepository.save(deck);
        });
        return MembyteReviewResultDto.builder()
                .nextReview(schedule.getDueAt())
                .nextReviewLabel(fsrsService.formatInterval(schedule.getDueAt(), now))
                .difficulty(schedule.getDifficulty())
                .stability(schedule.getStability())
                .build();
    }

    private MembyteDeckSummaryDto toDeckSummary(MembyteDeck deck, List<MembyteFlashcard> cards, Instant now) {
        return MembyteDeckSummaryDto.builder()
                .id(deck.getId())
                .title(deck.getTitle())
                .titleVi(deck.getTitleVi())
                .domain(deck.getDomain())
                .totalCards(cards.size())
                .dueCards((int) cards.stream().filter(card -> isDue(card, now)).count())
                .newCards((int) cards.stream().filter(MembyteFlashcard::isNew).count())
                .build();
    }

    private MembyteReviewCardDto toReviewCard(MembyteFlashcard card, Instant now) {
        Map<String, String> previews = new LinkedHashMap<>();
        for (MembyteReviewRatingRequest.Rating rating : MembyteReviewRatingRequest.Rating.values()) {
            FsrsService.Schedule schedule = fsrsService.preview(card, rating, now);
            previews.put(rating.name(), fsrsService.formatInterval(schedule.getDueAt(), now));
        }
        return MembyteReviewCardDto.builder()
                .id(card.getId())
                .deckId(card.getDeckId())
                .korean(card.getKorean())
                .pronunciation(card.getPronunciation())
                .vietnamese(card.getVietnamese())
                .english(card.getEnglish())
                .definitionEn(card.getDefinitionEn())
                .exampleKo(card.getExampleKo())
                .grammarNote(card.getGrammarNote())
                .context(card.getContext())
                .codeSnippet(card.getCodeSnippet())
                .audioUrl(card.getAudioUrl())
                .nextReviewTimes(previews)
                .build();
    }

    private boolean isDue(MembyteFlashcard card, Instant now) {
        return !card.isNew() && card.getDueAt() != null && !card.getDueAt().isAfter(now);
    }

    private VideoTokenMatch findVideoToken(Video video, String surface) {
        if (video.getKoreanSubtitles() == null) return null;
        for (Video.SubtitleLine subtitle : video.getKoreanSubtitles()) {
            if (subtitle.getTokens() == null) continue;
            for (Video.TokenizedWord token : subtitle.getTokens()) {
                if (surface.equals(token.getSurface())) {
                    return new VideoTokenMatch(token, subtitle.getText());
                }
            }
        }
        return null;
    }

    private void applyVideoTokenDetails(
            MembyteFlashcard card,
            Video.TokenizedWord token,
            String subtitleText,
            String savedWord) {
        card.setVocabularyId(savedWord);
        card.setKorean(savedWord);
        card.setPronunciation(token.getPronunciation());
        card.setVietnamese(token.getMeaningVi());
        card.setEnglish(token.getMeaningEn());
        card.setDefinitionEn(token.getDefinitionEn());
        card.setExampleKo(token.getExampleKo());
        card.setGrammarNote(token.getGrammarNote());
        card.setContext(subtitleText);
    }

    private void enrichVideoFlashcards(List<MembyteFlashcard> cards) {
        for (MembyteFlashcard card : cards) {
            if (card.getLessonId() == null || !card.getLessonId().startsWith("video:")) continue;
            String videoId = card.getLessonId().substring("video:".length());
            Video video = videoRepository.findById(videoId).orElse(null);
            if (video == null || video.getKoreanSubtitles() == null) continue;
            VideoTokenMatch match = findVideoTokenForCard(video, card);
            if (match == null) continue;
            Video.TokenizedWord token = match.token();
            String savedWord = token.getStem() == null || token.getStem().isBlank()
                    ? token.getSurface()
                    : token.getStem();
            card.setKorean(savedWord);
            card.setPronunciation(token.getPronunciation());
            card.setVietnamese(token.getMeaningVi());
            card.setEnglish(token.getMeaningEn());
            card.setDefinitionEn(token.getDefinitionEn());
            card.setExampleKo(token.getExampleKo());
            card.setGrammarNote(token.getGrammarNote());
            card.setContext(match.subtitleText());
            flashcardRepository.save(card);
        }
    }

    private VideoTokenMatch findVideoTokenForCard(Video video, MembyteFlashcard card) {
        for (Video.SubtitleLine subtitle : video.getKoreanSubtitles()) {
            if (subtitle.getTokens() == null) continue;
            for (Video.TokenizedWord token : subtitle.getTokens()) {
                if (card.getVocabularyId().equals(token.getSurface())
                        || card.getVocabularyId().equals(token.getStem())
                        || card.getKorean().equals(token.getSurface())
                        || card.getKorean().equals(token.getStem())) {
                    return new VideoTokenMatch(token, subtitle.getText());
                }
            }
        }
        return null;
    }

    private record VideoTokenMatch(Video.TokenizedWord token, String subtitleText) {
    }

    private String joinLabels(String topicLabel, String lessonLabel) {
        List<String> labels = new ArrayList<>();
        if (topicLabel != null && !topicLabel.isBlank()) labels.add(topicLabel);
        if (lessonLabel != null && !lessonLabel.isBlank()) labels.add(lessonLabel);
        return String.join(" · ", labels);
    }
}

File: kapor_flutter/lib/features/membyte/data/membyte_service.dart
import 'package:dio/dio.dart';

import '../../../core/network/api_client.dart';

class MemByteDeck {
  final String id;
  final String title;
  final String titleVi;
  final String domain;
  final int totalCards;
  final int dueCards;
  final int newCards;

  const MemByteDeck({
    required this.id,
    required this.title,
    required this.titleVi,
    required this.domain,
    required this.totalCards,
    required this.dueCards,
    required this.newCards,
  });

  factory MemByteDeck.fromJson(Map<String, dynamic> json) => MemByteDeck(
    id: json['id']?.toString() ?? '',
    title: json['title']?.toString() ?? '',
    titleVi: json['titleVi']?.toString() ?? '',
    domain: json['domain']?.toString() ?? '',
    totalCards: (json['totalCards'] as num?)?.toInt() ?? 0,
    dueCards: (json['dueCards'] as num?)?.toInt() ?? 0,
    newCards: (json['newCards'] as num?)?.toInt() ?? 0,
  );
}

class MemByteReviewSummary {
  final int totalDueCards;
  final int decksWithDueCards;

  const MemByteReviewSummary({
    required this.totalDueCards,
    required this.decksWithDueCards,
  });

  factory MemByteReviewSummary.fromJson(Map<String, dynamic> json) =>
      MemByteReviewSummary(
        totalDueCards: (json['totalDueCards'] as num?)?.toInt() ?? 0,
        decksWithDueCards: (json['decksWithDueCards'] as num?)?.toInt() ?? 0,
      );
}

class MemByteReviewCard {
  final String id;
  final String deckId;
  final String korean;
  final String pronunciation;
  final String vietnamese;
  final String english;
  final String definitionEn;
  final String exampleKo;
  final String grammarNote;
  final String context;
  final String codeSnippet;
  final String audioUrl;
  final Map<String, String> nextReviewTimes;

  const MemByteReviewCard({
    required this.id,
    required this.deckId,
    required this.korean,
    required this.pronunciation,
    required this.vietnamese,
    required this.english,
    required this.definitionEn,
    required this.exampleKo,
    required this.grammarNote,
    required this.context,
    required this.codeSnippet,
    required this.audioUrl,
    required this.nextReviewTimes,
  });

  factory MemByteReviewCard.fromJson(Map<String, dynamic> json) {
    final rawTimes = json['nextReviewTimes'];
    return MemByteReviewCard(
      id: json['id']?.toString() ?? '',
      deckId: json['deckId']?.toString() ?? '',
      korean: json['korean']?.toString() ?? '',
      pronunciation: json['pronunciation']?.toString() ?? '',
      vietnamese: json['vietnamese']?.toString() ?? '',
      english: json['english']?.toString() ?? '',
      definitionEn: json['definitionEn']?.toString() ?? '',
      exampleKo: json['exampleKo']?.toString() ?? '',
      grammarNote: json['grammarNote']?.toString() ?? '',
      context: json['context']?.toString() ?? '',
      codeSnippet: json['codeSnippet']?.toString() ?? '',
      audioUrl: json['audioUrl']?.toString() ?? '',
      nextReviewTimes: rawTimes is Map
          ? rawTimes.map(
              (key, value) => MapEntry(key.toString(), value.toString()),
            )
          : const {},
    );
  }
}

class MemByteService {
  final Dio _dio = ApiClient().dio;

  Future<void> saveVocabulary({
    required String lessonId,
    required String vocabularyId,
  }) async {
    try {
      await _dio.post('/membyte/lessons/$lessonId/flashcards/$vocabularyId');
    } on DioException catch (error) {
      throw Exception(
        _messageFromError(error, 'Không thể thêm thẻ vào MemByte.'),
      );
    }
  }

  Future<Set<String>> getSavedVocabularyIds(String lessonId) async {
    try {
      final data = _unwrap(
        await _dio.get('/membyte/lessons/$lessonId/saved-flashcards'),
      );
      if (data is! Map) {
        throw const FormatException('Dữ liệu flashcard đã lưu không hợp lệ.');
      }
      final vocabularyIds = data['vocabularyIds'];
      if (vocabularyIds is! List) return const {};
      return vocabularyIds.map((id) => id.toString()).toSet();
    } on DioException catch (error) {
      throw Exception(_messageFromError(error, 'Không thể tải thẻ đã lưu.'));
    }
  }

  Future<List<MemByteDeck>> getDecks() async {
    try {
      final data = _unwrap(await _dio.get('/membyte/decks'));
      if (data is! List) {
        throw const FormatException('Dữ liệu bộ thẻ không hợp lệ.');
      }
      return data
          .whereType<Map>()
          .map((item) => MemByteDeck.fromJson(Map<String, dynamic>.from(item)))
          .toList();
    } on DioException catch (error) {
      throw Exception(_messageFromError(error, 'Không thể tải MemByte.'));
    }
  }

  Future<MemByteReviewSummary> getReviewSummary() async {
    try {
      final data = _unwrap(await _dio.get('/membyte/review/summary'));
      if (data is! Map) {
        throw const FormatException('Dữ liệu tổng ôn không hợp lệ.');
      }
      return MemByteReviewSummary.fromJson(Map<String, dynamic>.from(data));
    } on DioException catch (error) {
      throw Exception(_messageFromError(error, 'Không thể tải tổng ôn.'));
    }
  }

  Future<List<MemByteReviewCard>> getReviewCards({
    String? deckId,
    required bool includeNew,
  }) async {
    try {
      final data = _unwrap(
        await _dio.get(
          '/membyte/review/cards',
          queryParameters: {
            if (deckId != null && deckId != 'all') 'deckId': deckId,
            'includeNew': includeNew,
            'limit': 50,
          },
        ),
      );
      if (data is! List) {
        throw const FormatException('Dữ liệu thẻ ôn không hợp lệ.');
      }
      return data
          .whereType<Map>()
          .map(
            (item) =>
                MemByteReviewCard.fromJson(Map<String, dynamic>.from(item)),
          )
          .toList();
    } on DioException catch (error) {
      throw Exception(_messageFromError(error, 'Không thể tải phiên ôn tập.'));
    }
  }

  Future<void> rateCard({
    required String cardId,
    required String rating,
    required int responseTimeMs,
  }) async {
    try {
      await _dio.post(
        '/membyte/review/rate',
        data: {
          'cardId': cardId,
          'rating': rating,
          'responseTimeMs': responseTimeMs,
        },
      );
    } on DioException catch (error) {
      throw Exception(
        _messageFromError(error, 'Không thể lưu kết quả ôn tập.'),
      );
    }
  }

  dynamic _unwrap(Response<dynamic> response) {
    final body = response.data;
    if (body is! Map<String, dynamic> || body['success'] != true) {
      final message = body is Map ? body['message'] : null;
      throw Exception(message ?? 'Phản hồi MemByte không hợp lệ.');
    }
    return body['data'];
  }

  String _messageFromError(DioException error, String fallback) {
    final data = error.response?.data;
    if (data is Map && data['message'] is String) {
      return data['message'] as String;
    }
    return fallback;
  }
}
